### PR TITLE
feat: OAuth2 endpoint audience

### PR DESCRIPTION
**Pull request by ndr-brt**
## What this PR changes/adds

Introduces a new "endpoing audience" setting value that will be used for validation, while the "provider audience" will be used on token request

## Why it does that

To permit more finer token claims validation

## Further notes

- moved `CredentialsRequestAdditionalParametersProvider` to `oauth2-spi`
- introduced `Oauth2DefaultServicesExtension`
- split `Oauth2ValidationRule` in 3 smaller rules to improve readability and tests

## Linked Issue(s)

Closes # <-- _insert Issue number if one exists_

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [x] assigned appropriate label? (exclude from changelog with label `no-changelog`)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/common/iam/oauth2/oauth2-core/src/test/java/org/eclipse/dataspaceconnector/iam/oauth2/core/identity/Oauth2ServiceImplTest.java [hunk 1]
```diff
@@ -211,9 +214,9 @@ void verifyExpiredToken() {
 
     @Test
     void verifyValidJwt() {
-        var jwt = createJwt(PROVIDER_AUDIENCE, Date.from(now.minusSeconds(1000)), new Date(System.currentTimeMillis() + 1000000));
+        var jwt = createJwt(ENDPOINT_AUDIENCE, Date.from(now.minusSeconds(1000)), new Date(System.currentTimeMillis() + 1000000));
```
Comment (ronjaquensel) [inline extensions/common/iam/oauth2/oauth2-core/src/test/java/org/eclipse/dataspaceconnector/iam/oauth2/core/identity/Oauth2ServiceImplTest.java:217]: Just a nit, but could we use the new endpoint audience in all test methods that verify a JWT (except for the ones testing for invalid audience of course)? Just to have it more consistent

**Review by ronjaquensel — APPROVED** [review]

File: extensions/common/iam/oauth2/oauth2-core/src/test/java/org/eclipse/dataspaceconnector/iam/oauth2/core/identity/Oauth2ServiceImplTest.java (hunk 1 above)
Comment (ndr-brt) [reply extensions/common/iam/oauth2/oauth2-core/src/test/java/org/eclipse/dataspaceconnector/iam/oauth2/core/identity/Oauth2ServiceImplTest.java:217]: done

File: extensions/common/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/rule/Oauth2ExpirationIssuedAtValidationRule.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            } else if (now.isBefore(convertToUtcTime(issuedAt))) {
+                return Result.failure("Current date/time before issued at (iat) claim in token");
+            }
+        }
+
+        return Result.success();
+    }
+
+    private static Instant convertToUtcTime(Date date) {
+        return ZonedDateTime.ofInstant(date.toInstant(), UTC).toInstant();
```
Comment (paullatzelsperger) [inline extensions/common/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/rule/Oauth2ExpirationIssuedAtValidationRule.java:66]: Micro nit: `OffsetDateTime` seems like a slightly more (semantically) suitable variant, but the actual difference seems to be minimal. For reference: [baeldung](https://www.baeldung.com/java-zoneddatetime-offsetdatetime) and [this link](https://simplesolution.dev/java-convert-date-to-offsetdatetime-in-utc/)

File: extensions/common/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/rule/Oauth2AudienceValidationRule.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import org.eclipse.dataspaceconnector.spi.result.Result;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.eclipse.dataspaceconnector.spi.jwt.JwtRegisteredClaimNames.AUDIENCE;
```
Comment (paullatzelsperger) [inline extensions/common/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/rule/Oauth2AudienceValidationRule.java:29]: Some documentation would be useful

File: extensions/common/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/rule/Oauth2ExpirationIssuedAtValidationRule.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.util.Date;
+import java.util.Map;
+
+import static java.time.ZoneOffset.UTC;
+import static org.eclipse.dataspaceconnector.spi.jwt.JwtRegisteredClaimNames.EXPIRATION_TIME;
+import static org.eclipse.dataspaceconnector.spi.jwt.JwtRegisteredClaimNames.ISSUED_AT;
+
+public class Oauth2ExpirationIssuedAtValidationRule implements TokenValidationRule {
```
Comment (paullatzelsperger) [inline extensions/common/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/rule/Oauth2ExpirationIssuedAtValidationRule.java:34]: Some documentation would be useful

File: extensions/common/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/rule/Oauth2NotBeforeValidationRule.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import java.time.Clock;
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.util.Date;
+import java.util.Map;
+
+import static java.time.ZoneOffset.UTC;
+import static org.eclipse.dataspaceconnector.spi.jwt.JwtRegisteredClaimNames.NOT_BEFORE;
+
+public class Oauth2NotBeforeValidationRule implements TokenValidationRule {
```
Comment (paullatzelsperger) [inline extensions/common/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/rule/Oauth2NotBeforeValidationRule.java:33]: Some documentation would be useful

**Review by paullatzelsperger — APPROVED** [review]

**Review by denisneuling — APPROVED** [review]

**Pull request merged**